authz/rbac_translator.go

```diff
@@ -0,0 +1,310 @@
+	default:
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_ExactMatch{ExactMatch: value},
+		}
+	}
+}
+
+func parsePrincipalNames(principalNames []string) []*v3rbacpb.Principal {
+	var ps []*v3rbacpb.Principal
+	for _, principalName := range principalNames {
+		newPrincipalName := &v3rbacpb.Principal{
+			Identifier: &v3rbacpb.Principal_Authenticated_{
+				Authenticated: &v3rbacpb.Principal_Authenticated{
+					PrincipalName: getStringMatcher(principalName),
+				},
+			}}
+		ps = append(ps, newPrincipalName)
+	}
+	return ps
+}
+
+func parsePeer(source peer) (*v3rbacpb.Principal, error) {
+	if source.Principals != nil {
+		if len(source.Principals) == 0 {
+			return nil, fmt.Errorf(`"principals" is empty`)
+		}
+		return principalOr(parsePrincipalNames(source.Principals)), nil
+	}
+	return &v3rbacpb.Principal{
+		Identifier: &v3rbacpb.Principal_Any{
+			Any: true,
+		},
+	}, nil
+}
+
+func parsePaths(paths []string) []*v3rbacpb.Permission {
+	var ps []*v3rbacpb.Permission
+	for _, path := range paths {
+		newPath := &v3rbacpb.Permission{
+			Rule: &v3rbacpb.Permission_UrlPath{
+				UrlPath: &v3matcherpb.PathMatcher{
+					Rule: &v3matcherpb.PathMatcher_Path{Path: getStringMatcher(path)}}}}
+		ps = append(ps, newPath)
+	}
+	return ps
+}
+
+func parseHeaderValues(key string, values []string) []*v3rbacpb.Permission {
+	var vs []*v3rbacpb.Permission
+	for _, value := range values {
+		newHeader := &v3rbacpb.Permission{
+			Rule: &v3rbacpb.Permission_Header{
+				Header: getHeaderMatcher(key, value)}}
+		vs = append(vs, newHeader)
+	}
+	return vs
+}
+
+var unsupportedHeaders = map[string]bool{
+	"host":                true,
+	"connection":          true,
+	"keep-alive":          true,
+	"proxy-authenticate":  true,
+	"proxy-authorization": true,
+	"te":                  true,
+	"trailer":             true,
+	"transfer-encoding":   true,
+	"upgrade":             true,
+}
+
+func unsupportedHeader(key string) bool {
+	return strings.HasPrefix(key, ":") || strings.HasPrefix(key, "grpc-") || unsupportedHeaders[key]
+}
+
+func parseHeaders(headers []header) ([]*v3rbacpb.Permission, error) {
+	var hs []*v3rbacpb.Permission
+	for i, header := range headers {
+		if header.Key == "" {
+			return nil, fmt.Errorf(`"headers" %d: "key" is not present`, i)
+		}
+		header.Key = strings.ToLower(header.Key)
+		if unsupportedHeader(header.Key) {
+			return nil, fmt.Errorf(`"headers" %d: unsupported "key" %s`, i, header.Key)
+		}
+		if header.Values == nil || len(header.Values) == 0 {
+			return nil, fmt.Errorf(`"headers" %d: "values" is not present/empty`, i)
+		}
+		values := parseHeaderValues(header.Key, header.Values)
+		hs = append(hs, permissionOr(values))
+	}
+	return hs, nil
+}
+
+func parseRequest(request request) (*v3rbacpb.Permission, error) {
+	var and []*v3rbacpb.Permission
+	if request.Paths != nil {
+		if len(request.Paths) == 0 {
+			return nil, fmt.Errorf(`"paths" is empty`)
+		}
```

Comment:
Wait, this confuses me.

Is it okay for `Paths` to be `nil`?  Why is it not okay for it to be empty but non-`nil` if so?

If it's not okay for it to be `nil` then this doesn't check that.

I _think_ what you meant was:

```go
if len(request.Paths) == 0 || len(request.Headers) == 0 { return some error }
...
```

But if that's not right, then you should do:

```go
if len(request.Paths) > 0 {
	and = append(...)
}
if len(request.Headers) > 0 {
	and = append(...)
}
```

Reply:
paths, headers and principals are optional fields. Hence it is okay for them to be nil.

Consider the following policies
1.
{
  "name": "authz..",
  "allow_rules": [{
    "name":"...",
   "request": {"headers":[{"key"....}]},
  }]
}

2.
{
  "name": "authz..",
  "allow_rules": [{
    "name":"...",
   "request": {"paths":[]},
  }]
}

With current code, (2) would error out and (1) would translate successfully.
If we don't do the nil check, then (1) would fail, which is unintended behavior.

if we removed the error on empty paths, and appended the policy if paths has entries, then now (2) would translate successfully to RBAC ALLOW policy permission ANY.

I think the question is should (2) error out or translate successfully. I am fine either way. Wdyt?

Reply:
> if we removed the error on empty paths, and appended the policy if paths has entries, then now (2) would translate successfully to RBAC ALLOW policy permission ANY.

Unless it's semantically useful, we should probably do the same thing when a field is empty vs. absent.  But we also need to be consistent within languages.  This should be covered in the gRFC.

Reply:
Sounds good. Will mention in gRFC and ensure consistent behavior across languages.
Updated the code, to translate (2) successfully.